File: packages/javascript/examples/websocket-example.ts
/**
 * Tavo AI JavaScript SDK - WebSocket Example
 *
 * This example demonstrates how to use WebSocket connections for real-time features
 * including scan progress updates, notifications, and general broadcasts.
 */

import { GeneralMessage, NotificationMessage, ScanUpdateMessage, TavoClient } from '../src/index.js';

// Initialize the client
const client = new TavoClient({
    jwtToken: 'your-jwt-token-here', // or apiKey/sessionToken
    baseURL: 'https://api.tavoai.net',
});

// Example 1: Connect to scan progress updates
async function monitorScanProgress(scanId: string) {
    try {
        const connectionId = await client.websocket.connectToScan(
            scanId,
            (message: ScanUpdateMessage) => {
                console.log('Scan update:', message);

                switch (message.update_type) {
                    case 'started':
                        console.log(`Scan ${message.scan_id} has started`);
                        break;
                    case 'progress':
                        console.log(`Scan progress: ${message.message}`);
                        break;
                    case 'completed':
                        console.log(`Scan ${message.scan_id} completed successfully`);
                        // Disconnect when scan is done
                        client.websocket.disconnect(connectionId);
                        break;
                    case 'error':
                        console.error(`Scan error: ${message.message}`);
                        client.websocket.disconnect(connectionId);
                        break;
                }
            }
        );

        console.log(`Connected to scan ${scanId} with connection ID: ${connectionId}`);

        // Optional: Send a ping message
        setTimeout(() => {
            client.websocket.sendMessage(connectionId, { type: 'ping' });
        }, 5000);

    } catch (error) {
        console.error('Failed to connect to scan WebSocket:', error);
    }
}

// Example 2: Connect to notifications
async function monitorNotifications() {
    try {
        const connectionId = await client.websocket.connectToNotifications(
            (message: NotificationMessage) => {
                console.log('Notification received:', message);

                switch (message.type) {
                    case 'info':
                        console.info(message.title, message.message);
                        break;
                    case 'warning':
                        console.warn(message.title, message.message);
                        break;
                    case 'error':
                        console.error(message.title, message.message);
                        break;
                    case 'success':
                        console.log('✅', message.title, message.message);
                        break;
                }
            }
        );

        console.log(`Connected to notifications with connection ID: ${connectionId}`);

    } catch (error) {
        console.error('Failed to connect to notifications WebSocket:', error);
    }
}

// Example 3: Connect to general broadcasts
async function monitorGeneralBroadcasts() {
    try {
        const connectionId = await client.websocket.connectToGeneral(
            (message: GeneralMessage) => {
                console.log('General broadcast:', message);
            }
        );

        console.log(`Connected to general broadcasts with connection ID: ${connectionId}`);

    } catch (error) {
        console.error('Failed to connect to general WebSocket:', error);
    }
}

// Example 4: Clean up connections
async function cleanup() {
    console.log('Disconnecting all WebSocket connections...');
    await client.websocket.disconnectAll();
    console.log('All connections closed');
}

// Usage examples
async function main() {
    // Start monitoring notifications (long-running)
    await monitorNotifications();

    // Start monitoring general broadcasts (long-running)
    await monitorGeneralBroadcasts();

    // Create and monitor a scan
    try {
        const scanResult = await client.scans.create({
            repositoryUrl: 'https://github.com/example/repo',
            scanType: 'security'
        });

        console.log('Scan created:', scanResult);
        await monitorScanProgress(scanResult.id);

    } catch (error) {
        console.error('Scan creation failed:', error);
    }

    // Keep the process running for a while to receive messages
    console.log('Listening for real-time updates... Press Ctrl+C to exit');

    // In a real application, you might want to handle graceful shutdown
    process.on('SIGINT', async () => {
        await cleanup();
        process.exit(0);
    });
}

// Run the example
main().catch(console.error);

File: packages/python/examples/websocket_example.py
#!/usr/bin/env python3
"""
Tavo AI Python SDK - WebSocket Example

This example demonstrates how to use WebSocket connections for real-time features
including scan progress updates, notifications, and general broadcasts.
"""

import asyncio
import os
from tavo.client import (
    TavoClient,
    ScanUpdateMessage,
    NotificationMessage,
    GeneralMessage,
)


async def monitor_scan_progress(client: TavoClient, scan_id: str):
    """Monitor scan progress via WebSocket"""
    try:
        connection_id = await client.websocket().connect_to_scan(
            scan_id, lambda message: handle_scan_message(message)
        )

        print(f"Connected to scan {scan_id} with connection ID: {connection_id}")

        # Optional: Send a ping message
        await asyncio.sleep(5)
        await client.websocket().send_message(connection_id, {"type": "ping"})

        # Keep connection alive for demonstration
        await asyncio.sleep(300)  # 5 minutes

    except Exception as error:
        print(f"Failed to connect to scan WebSocket: {error}")


def handle_scan_message(message: ScanUpdateMessage):
    """Handle scan update messages"""
    print(f"Scan update: {message}")

    if message.update_type == "started":
        print(f"Scan {message.scan_id} has started")
    elif message.update_type == "progress":
        print(f"Scan progress: {message.message}")
    elif message.update_type == "completed":
        print(f"Scan {message.scan_id} completed successfully")
    elif message.update_type == "error":
        print(f"Scan error: {message.message}")


async def monitor_notifications(client: TavoClient):
    """Monitor user notifications via WebSocket"""
    try:
        connection_id = await client.websocket().connect_to_notifications(
            lambda message: handle_notification(message)
        )

        print(f"Connected to notifications with connection ID: {connection_id}")

        # Keep connection alive
        await asyncio.sleep(300)  # 5 minutes

    except Exception as error:
        print(f"Failed to connect to notifications WebSocket: {error}")


def handle_notification(message: NotificationMessage):
    """Handle notification messages"""
    print(f"Notification received: {message}")

    if message.type == "info":
        print(f"ℹ️  {message.title}: {message.message}")
    elif message.type == "warning":
        print(f"⚠️  {message.title}: {message.message}")
    elif message.type == "error":
        print(f"❌ {message.title}: {message.message}")
    elif message.type == "success":
        print(f"✅ {message.title}: {message.message}")


async def monitor_general_broadcasts(client: TavoClient):
    """Monitor general broadcasts via WebSocket"""
    try:
        connection_id = await client.websocket().connect_to_general(
            lambda message: handle_general_message(message)
        )

        print(f"Connected to general broadcasts with connection ID: {connection_id}")

        # Keep connection alive
        await asyncio.sleep(300)  # 5 minutes

    except Exception as error:
        print(f"Failed to connect to general WebSocket: {error}")


def handle_general_message(message: GeneralMessage):
    """Handle general broadcast messages"""
    print(f"General broadcast: {message}")


async def cleanup_connections(client: TavoClient):
    """Clean up all WebSocket connections"""
    print("Disconnecting all WebSocket connections...")
    await client.websocket().disconnect_all()
    print("All connections closed")


async def main():
    """Main example function"""
    # Initialize the client
    client = TavoClient(
        jwt_token=os.getenv("TAVO_JWT_TOKEN"),  # or api_key/session_token
        base_url="https://api.tavoai.net",
    )

    try:
        # Start monitoring notifications (in background)
        _ = asyncio.create_task(monitor_notifications(client))

        # Start monitoring general broadcasts (in background)
        _ = asyncio.create_task(monitor_general_broadcasts(client))

        # Create and monitor a scan
        try:
            scan_result = await client.scans().create(
                repository_url="https://github.com/example/repo", scan_type="security"
            )

            print(f"Scan created: {scan_result}")
            await monitor_scan_progress(client, scan_result["id"])

        except Exception as error:
            print(f"Scan creation failed: {error}")

    except KeyboardInterrupt:
        print("\nShutting down...")
    finally:
        await cleanup_connections(client)


if __name__ == "__main__":
    print("Tavo AI WebSocket Example")
    print("Make sure to set TAVO_JWT_TOKEN environment variable")
    print("Press Ctrl+C to exit\n")

    asyncio.run(main())
